miniapps/nurbs/nurbs_curveint.cpp

```diff
@@ -0,0 +1,238 @@
+//               B-splines, and curves with not all weights being 1, NURBS. The
+//               spacing in both cases is chosen differently.
+
+
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace std;
+using namespace mfem;
+
+KnotVector *UniformKnotVector(int order, int ncp)
+{
+   if (order>=ncp)
+   {
+      mfem_error("UniformKnotVector: ncp should be at least order + 1");
+   }
+   KnotVector *kv = new KnotVector(order, ncp);
+
+   for (int i = 0; i < order+1; i++)
+   {
+      (*kv)[i] = 0.0;
+   }
+   for (int i = order+1; i < ncp; i++)
+   {
+      (*kv)[i] = (i-order)/double(ncp-order);
+   }
+   for (int i = ncp ; i < ncp + order + 1; i++)
+   {
+      (*kv)[i] = 1.0;
+   }
+   return kv;
+}
+
+
+
+int main(int argc, char *argv[])
+{
+   // Parse command-line options.
+   OptionsParser args(argc, argv);
+
+   double l       = 1.0;
+   double a       = 0.1;
+   int ncp        = 9;
+   int order      = 2;
+   bool ifbspline = true;
+
+   args.AddOption(&l, "-l", "--box-side-length",
+                  "Height and width of the box");
+   args.AddOption(&a, "-a", "--sine-ampl",
+                  "Amplitude of the fitted sine function.");
+   args.AddOption(&a, "-a", "--sine-ampl",
+                  "Amplitude of the fitted sine function.");
+   args.AddOption(&ncp, "-n", "--ncp",
+                  "Number of control points used over four box sides.");
+   args.AddOption(&order, "-o", "--order",
+                  "Order of the NURBSPatch");
+   args.AddOption(&ifbspline, "-uw", "--unit-weight", "-nw",
+                  "--non-unit-weight",
+                  "Use a weight of one only for the NURBS-patches (default) or not: B-splines");
+
+   // Parse and print commandline options
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
+   }
+   args.PrintOptions(cout);
+
+   if (order < 2 && ifbspline == false)
+   {
+      mfem_error("For a non unity weight, the order should be at least 2.");
+   }
+
+   KnotVector *kv_o1 = UniformKnotVector(1, 2);
+   KnotVector *kv    = UniformKnotVector(order, ncp);
+
+   // 1. Create a box shaped NURBS patch
+   NURBSPatch patch(kv_o1, kv_o1, 3);
+
+   // Set weights
+   for (int j = 0; j < 2; j++)
+      for (int i = 0; i < 2; i++)
+      {
+         patch(i,j,2) = 1.0;
+      }
+
+   // Define patch corners which are box corners
+   patch(0,0,0) = -0.5*l;
+   patch(0,0,1) = -0.5*l;
+
+   patch(1,0,0) = 0.5*l;
+   patch(1,0,1) = -0.5*l;
+
+
+   patch(0,1,0) = -0.5*l;
+   patch(0,1,1) = 0.5*l;
+
+   patch(1,1,0) = 0.5*l;
+   patch(1,1,1) = 0.5*l;
+
+   // 2. Interpolation process
+   Array<Vector*> xy(2);
+   xy[0] = new Vector();
+   xy[1] = new Vector();
+   Vector xi_args, u_args;
+   Array<int> i_args;
+   xy[0]->SetSize(ncp); xy[1]->SetSize(ncp);
+
+   // Refine direction which has fitting
+   patch.DegreeElevate(0, order-kv_o1->GetOrder());
+   if (!ifbspline)
+   {
+      patch(1,0,2) = sqrt(2)/2;
```

Comment:
Why should only this one point get the non-unit weight? Depending on the order and number of control points, there can be many points on the bottom of the square. Some documentation explaining the choice of weights would be helpful. Also, the result is strange in the case `nurbs_curveint -nw -n 4 -o 3`.

Reply:
_Depending on the order and number of control points, there can be many points on the bottom of the square. Some documentation explaining the choice of weights would be helpful.  _
I wrote the code with second order patches in mind and did not realize it could have a purpose for higher order patches. I have improved the code in [25f8deb](https://github.com/mfem/mfem/pull/3232/commits/25f8deb432b4e8364a0609f36323f356e398a940) and [20eb64d](https://github.com/mfem/mfem/pull/3232/commits/20eb64d8cf050879baf67303136f404d9895be19). It contains an explanation for the choice of weights and the application of the weight is improved. I have also improved the curve interpolation procedure and clarified the conversion from carthesian coordinates to homogeneous coordinates.

_Also, the result is strange in the case nurbs_curveint -nw -n 4 -o 3_
This had several reasons. 1) As stated above I have improved the curve interpolation procedure, some part of the procedure made it very messy. 2) Apparantly FindMaxima did not work correctly for `KnotVectors` with one element. For lower orders we just have been lucky it worked out. I have applied a quick fix. But I will have to see if this is the one we need.

Reply:
And most important. Thank your for pointing it out!

Reply:
Thank you @JacobLotz. This is a very nice example, and the first of its kind in MFEM. The NURBS features in MFEM really needed some examples like this, and I appreciate your contributions. Please update us when you are sure the fix is sufficient, and I will review again.

Reply:
@dylan-copeland 
I have tried applying Newtons method in finding the maxima of the shapefunctions of the knotvectors as you suggested in an other comment. This was a bit tricky as the Newtons method would find a minimum instead of a maximum sometimes. I think I can push Newton's method in the right direction but it would not increase the elegance of the code in my opinion. 

I think the current approach is better as we are very sure we will find the maximum. I have disturbed the initial guess of both `arg1` and `arg2` with a very small number `1e-16` (very close to machine precision) such that we are able to apply the function to `KnotVector`s with one element. With the initial choice of `arg1` and `arg2` without the disturbance we would also not find exactly 0 or 1 so the disturbance does not influence the results. 

I think we can do the review again.

Reply:
Thanks @JacobLotz for the improvements. The example is working well now.

I don't understand the issue that requires perturbing the initial guesses by 1e-16. I set `arg1=0` and `arg2=1` and ran with one element, without any trouble. Is there an example where it fails?

Reply:
@dylan-copeland, you are able to run it, but the answer is not correct. You can see it for instance by viewing the result in Visit and and increasing multires control or printing the locations of the maxima. Without the perturbation it is only able to find about halve of the locations of the maxima of the `KnotVector`. A good test would be for instance `./nurbs_curveint -n 6 -o 5`. I have attached two print screens of visit below and the locations of the maxima `xi`.  

In short: This results from evaluating the if statement at line 379 `if (max1 < max2)` where in the case of one element and without the disturbance `max1 = max2 = 0`. This will push the algorithm in the wrong direction such that it only finds about halve of the locations of the maxima.

I hope this clarifies the issue!

**Without the disturbance**
`xi = [2.77556e-17 0.25 0.4 0.5 0.5 1]`

![faulty](https://user-images.githubusercontent.com/69582468/200261782-d63000a5-83b0-4778-9f71-e8879ca80d12.png)

**With the disturbance**
`xi = [1.27756e-16 0.25 0.4 0.6 0.75 1]`

![correct](https://user-images.githubusercontent.com/69582468/200261840-d6701244-bec6-4f92-b439-c73d50ea434a.png)

Reply:
This does help to clarify the difficulty in solving for the maximum. Finding the maximum in an interval can be tricky, and I see now why this binary search type of algorithm needs to start slightly away from the endpoints, where the basis function can be exactly zero. This algorithm is acceptable for this PR, in my opinion, but maybe in the future it could be improved if someone wants to make it more robust and efficient. As we discussed already, Newton's method could be much faster, but it needs a sufficiently close initial guess. It seems that a small number of initial guesses could be used, perhaps chosen according to which basis function is used. Even better, I wonder whether it is possible to find the maxima analytically, based on the code in `KnotVector::CalcShape`. In any case, such optimizations could be done in the future.

Reply:
@dylan-copeland I agree. This is not the most elegant solution but works.
Keep in mind that robustness is more important then performance. These routines are typically only run a moderate amount of times when building a mesh.